src/Compose/ConcurrentTransitionManagerContainer.cs

```
@@ -1,88 +1,88 @@
-			while (!managers.IsEmpty)
-				if (managers.TryTake(out manager))
-					restoredManagers.Add(Restore(Helper.MakeGenericType(serviceType).GetTypeInfo(), manager));
-			while (!restoredManagers.IsEmpty)
-				if (restoredManagers.TryTake(out manager))
-					managers.Add(manager);
-		}
-
-		private object Restore(TypeInfo helper, object manager)
-		{
-			helper.DeclaredMethods.Single(x => x.Name == "Restore").Invoke(null, new[] { manager });
-			return manager;
-		}
-
-		public void Snapshot()
-		{
-			foreach (var kvp in Managers)
-				Snapshot(kvp.Key, kvp.Value);
-		}
-
-		private void Snapshot(Type serviceType, ConcurrentBag<object> managers)
-		{
-			var snapshottedManagers = new ConcurrentBag<object>();
-			object manager = null;
-			while (!managers.IsEmpty)
-				if (managers.TryTake(out manager))
-					snapshottedManagers.Add(Snapshot(Helper.MakeGenericType(serviceType).GetTypeInfo(), manager));
-			while (!snapshottedManagers.IsEmpty)
-				if (snapshottedManagers.TryTake(out manager))
-					managers.Add(manager);
-		}
-
-		private object Snapshot(TypeInfo helper, object manager)
-		{
-			helper.DeclaredMethods.Single(x => x.Name == "Snapshot").Invoke(null, new[] { manager });
-			return manager;
-		}
-	}
-}
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Reflection;
+
+namespace Compose
+{
+	internal sealed class ConcurrentTransitionManagerContainer : TransitionManagerContainer
+	{
+		private readonly ConcurrentDictionary<Type, ConcurrentBag<object>> Managers
+			= new ConcurrentDictionary<Type, ConcurrentBag<object>>();
+		private static readonly Type Helper = typeof(GenericHelper<>);
+
+		private static class GenericHelper<T>
+		{
+			private static readonly TypeInfo Disposable = typeof(IDisposable).GetTypeInfo();
+			public static void Restore(object untypedManager)
+			{
+				var manager = untypedManager as DynamicRegister<T>;
+				if (Disposable.IsAssignableFrom(manager.CurrentService.GetType().GetTypeInfo()))
```

I still feel that the arguments should be explicitly checked, in many cases I'd imagine that it would make no/little difference being internal and not as open to rubbish arguments being passed in, however there will be cases where it's more useful.

Argument 'theParameter' cannot be null. vs. later of the generic "object not set to an instance of an object", the more defensive and useful error message would be more preferential as a small nitpick - maybe that's just my preference (Yes, I know you can actually just read the rest of the exception info etc etc - but WHATEVER)?

Basically, I can't say you're wrong @smudge202, as tests should cover the scenarios - you're right, but still feel you should check the arguments explicitly :wink: at least you can fail early if required and as above provide a better error message if required than some of the horridly vague framework messages.

I've just re-read this comment and basically it's just waffle and haven't made a concrete argument for or against :disappointed: